File: shop_frontend/src/services/productService.js
import api from './api';

const productService = {
    //Get all products with pagination and filters
    getProducts: async ({ page = 0, size = 10, sort = 'name,asc', categoryId, animalId, search } = {}) => {
        try {
            const params = { page, size, sort };
            if (categoryId) params.categoryId = categoryId;
            if (animalId) params.animalId = animalId;
            if (search) params.search = search;

            const response = await api.get('/products', { params });
            return response.data;
        } catch (error) {
            throw error.response?.data || error.message;
        }
    },

    // Get all animals
    getAnimals: async () => {
        try {
            const response = await api.get('/animals');
            return response.data;
        } catch (error) {
            throw error.response?.data || error.message;
        }
    },

    // Get subcategories by parent ID
    getSubCategories: async (parentId) => {
        try {
            const response = await api.get(`/categories/${parentId}/subcategories`);
            return response.data;
        } catch (error) {
            throw error.response?.data || error.message;
        }
    },

    //Get product by ID
    getProductById: async (productId) => {
        try {
            const response = await api.get(`/products/${productId}`);
            return response.data;
        } catch (error) {
            throw error.response?.data || error.message;
        }
    },

    // Search products
    searchProducts: async (query) => {
        try {
            const response = await api.get('/products/search', {
                params: { q: query },
            });
            return response.data;
        } catch (error) {
            throw error.response?.data || error.message;
        }
    },

    //Get products by category
    getProductsByCategory: async (categoryId, page = 0, size = 10) => {
        try {
            const response = await api.get(`/products/category/${categoryId}`, {
                params: { page, size },
            });
            return response.data;
        } catch (error) {
            throw error.response?.data || error.message;
        }
    },


    // ADMIN - Create new product
    createProduct: async (productData) => {
        try {
            const response = await api.post('/products', productData);
            return response.data;
        } catch (error) {
            throw error.response?.data || error.message;
        }
    },

    // ADMIN - Update product
    updateProduct: async (productId, productData) => {
        try {
            const response = await api.put(`/products/${productId}`, productData);
            return response.data;
        } catch (error) {
            throw error.response?.data || error.message;
        }
    },
};

export default productService;

File: shop.backend/src/main/java/com/api/shop/backend/service/ProductService.java
package com.api.shop.backend.service;

import com.api.shop.backend.dto.CreateProductRequest;
import com.api.shop.backend.model.Category;
import com.api.shop.backend.model.Product;
import com.api.shop.backend.repository.CategoryRepository;
import com.api.shop.backend.repository.ProductRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import java.util.List;

@Service
public class ProductService {

    private final ProductRepository productRepository;
    private final CategoryRepository categoryRepository;
    private final com.api.shop.backend.repository.AnimalRepository animalRepository;

    @Autowired
    public ProductService(ProductRepository productRepository,
            CategoryRepository categoryRepository,
            com.api.shop.backend.repository.AnimalRepository animalRepository) {
        this.productRepository = productRepository;
        this.categoryRepository = categoryRepository;
        this.animalRepository = animalRepository;
    }

    public List<Product> searchProducts(String search) {
    if (search == null || search.trim().isEmpty()) {
        return getAllProducts(null, null);
    }
    return productRepository.searchProducts(search);
}

    public List<Product> getAllProducts(Long categoryId, Long animalId) {
        org.springframework.data.jpa.domain.Specification<Product> spec = (root, query, cb) -> {
            java.util.List<jakarta.persistence.criteria.Predicate> predicates = new java.util.ArrayList<>();

            if (categoryId != null) {
                jakarta.persistence.criteria.Join<Product, Category> categoryJoin = root.join("category",
                        jakarta.persistence.criteria.JoinType.LEFT);
                jakarta.persistence.criteria.Join<Category, Category> parentJoin = categoryJoin.join("parent",
                        jakarta.persistence.criteria.JoinType.LEFT);

                predicates.add(cb.or(
                        cb.equal(categoryJoin.get("id"), categoryId),
                        cb.equal(parentJoin.get("id"), categoryId)));
            }

            if (animalId != null) {
                predicates.add(cb.equal(root.get("animal").get("id"), animalId));
            }

            return cb.and(predicates.toArray(new jakarta.persistence.criteria.Predicate[0]));
        };

        return productRepository.findAll(spec);
    }

    public Product getProductById(Long id) {
        return productRepository.findById(id)
                .orElseThrow(() -> new RuntimeException("Product not found"));
    }

    public Product createProduct(CreateProductRequest request) {
        // Find category
        Category category = categoryRepository.findById(request.getCategoryId())
                .orElseThrow(() -> new RuntimeException("Category not found"));

        // Create new product
        Product product = new Product();
        product.setName(request.getName());
        product.setDescription(request.getDescription());
        product.setDescriptif(request.getDescriptif());
        product.setPrice(request.getPrice());
        product.setStock(request.getStock());
        product.setImageUrl(request.getImageUrl());
        product.setCategory(category);

        if (request.getAnimalId() != null) {
            com.api.shop.backend.model.Animal animal = animalRepository.findById(request.getAnimalId())
                    .orElseThrow(() -> new RuntimeException("Animal not found"));
            product.setAnimal(animal);
        }

        return productRepository.save(product);
    }

    public Product updateProduct(Long id, CreateProductRequest request) {
        Product product = getProductById(id);

        // Find category if changed
        Category category = categoryRepository.findById(request.getCategoryId())
                .orElseThrow(() -> new RuntimeException("Category not found"));

        // Update fields
        // Update fields
        product.setName(request.getName());
        product.setDescription(request.getDescription());
        product.setDescriptif(request.getDescriptif());
        product.setPrice(request.getPrice());
        product.setStock(request.getStock());
        product.setImageUrl(request.getImageUrl());
        product.setCategory(category);

        if (request.getAnimalId() != null) {
            com.api.shop.backend.model.Animal animal = animalRepository.findById(request.getAnimalId())
                    .orElseThrow(() -> new RuntimeException("Animal not found"));
            product.setAnimal(animal);
        } else {
            product.setAnimal(null);
        }

        return productRepository.save(product);
    }

    public void deleteProduct(Long id) {
        Product product = getProductById(id);
        productRepository.delete(product);
    }
}